packages/apps/ability-uniswap-swap/src/lib/ability-helpers/get-gas-params.ts

```diff
@@ -1,43 +1,117 @@
-import { ethers } from 'ethers';
+import { formatUnits, JsonRpcProvider } from 'ethers-v6';
 
-export interface GasParams {
-  estimatedGas: ethers.BigNumber;
-  maxFeePerGas: ethers.BigNumber;
-  maxPriorityFeePerGas: ethers.BigNumber;
+export type GasParams = LegacyGasParams | Eip1559GasParams;
+
+interface LegacyGasParams {
+  gasPrice: string; // in wei
+  estimatedGas: string;
 }
 
-export const getGasParams = async (
-  provider: ethers.providers.Provider,
-  estimatedGas: ethers.BigNumber,
-): Promise<GasParams> => {
-  // Fetch block and fee data
-  const [block, feeData] = await Promise.all([provider.getBlock('latest'), provider.getFeeData()]);
-  // Determine baseFeePerGas, with multiple fallbacks
-  let baseFeePerGas = block.baseFeePerGas;
-  if (baseFeePerGas == null) {
-    // Fallback to feeData.gasPrice if available
-    baseFeePerGas = feeData && feeData.gasPrice ? feeData.gasPrice : await provider.getGasPrice();
-  }
-  // Final fallback to a nominal value if still null
-  if (baseFeePerGas == null) {
-    baseFeePerGas = ethers.utils.parseUnits('1', 'gwei');
-  }
+interface Eip1559GasParams {
+  maxFeePerGas: string; // in wei
+  maxPriorityFeePerGas: string; // in wei
+  estimatedGas: string;
+}
+
+const DEFAULT_GAS_LIMIT_BUFFER = 50; // adding 50% buffer to the estimated gas
+const DEFAULT_BASE_FEE_PER_GAS_MULTIPLIER = 2; // maxFee = base * 2 + tip
+
+export const getGasParams = async ({
+  rpcUrl,
+  estimatedGas,
+  gasLimitBuffer = DEFAULT_GAS_LIMIT_BUFFER,
+  baseFeePerGasMultiplier = DEFAULT_BASE_FEE_PER_GAS_MULTIPLIER,
+}: {
+  rpcUrl: string;
+  estimatedGas: string;
+  gasLimitBuffer?: number;
+  baseFeePerGasMultiplier?: number;
+}): Promise<GasParams> => {
```

Comment:
Why don't we make this receive the `calldata`, `nonce`, etc and estimate gas like a normal tx?
`const gasLimit = await provider.estimateGas({ from, to, data, value, nonce });`

Seems a very complicated calculation still based on the faulty `estimatedGas` we got from uniswap sdk (ethers v5)

For the L2->L1 settlement, we can use the buffer or try to get that correctly with the oracle

Reply:
Simplified, we now just use ether's v6 `populateTransaction` and pass it:

```
to,
from,
value,
data,
rpcUrl,
chainId,
gasBufferPercentage,
baseFeePerGasBufferPercentage,
```